Packaged code for PyPI by lacerdar

iatzak opened the pull request:
<img width="958" alt="testPyPI" src="https://user-images.githubusercontent.com/112731474/199119155-eb618449-c49a-494d-8933-44ed2c03f24f.png">

IshaanDesai reviewed (commented):
I could not import the package in an interactive Python shell:

```bash
Installing collected packages: lacerdar_diffusion2d
Successfully installed lacerdar_diffusion2d-0.0.1
desaiin@lapus302:~$ python3
Python 3.10.6 (main, Nov  2 2022, 18:53:38) [GCC 11.3.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> from lacerdar_diffusion2d import diffusion2d
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
ModuleNotFoundError: No module named 'lacerdar_diffusion2d'
>>> import lacerdar_diffusion2d
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
ModuleNotFoundError: No module named 'lacerdar_diffusion2d'
>>> 
```

iatzak commented:
@IshaanDesai I firstly followed your instructions on the [exercise page](https://github.com/Simulation-Software-Engineering/Lecture-Material/blob/main/03_building_and_packaging/pypi_exercise.md) (now corrected after [issue 125](https://github.com/Simulation-Software-Engineering/Lecture-Material/issues/125)), which were saying to create a package with the name `<your-GitLab-username>-diffusion2d` . I did it that way, then tried to import the package and failed, because of the minus sign. I then saw in the [Ilias forum](https://ilias3.uni-stuttgart.de/ilias.php?ref_id=3117404&cmdClass=ilobjforumgui&thr_pk=156246&page=0&cmd=viewThread&cmdNode=z7:pe&baseClass=ilRepositoryGUI) that you gave a solution to this (name the package `<your-GitLab-username>_diffusion2d`). Therefore, I changed my code to include this. However, this is only a solution for those who didn't do the exercise before, since Test PyPI doesn't allow me to have a new package with a name that is "too similar" to a previous package. Thus you can try to import [my package from Test PyPI](https://test.pypi.org/project/lacerdar-diffusion2d/) with `pip install -i https://test.pypi.org/simple/ lacerdar-diffusion2d`, but it will result in the aforementioned error.

If you have a solution to this, just let me know and I'll do it. (perhaps change the package name to something "not so similar"?)

IshaanDesai commented:
@iatzak yes I understood that you went through the whole procedure of using the minus sign first and then the underscore. I also understand that TestPyPI would probably complain if you try to upload a package only with a changed symbol as it is too similar. One suggestion is to delete the previous package before uploading the new one. What surprised me a bit was that I could in fact install the package `lacerdar_diffusion2d` (have a look at the output I have added above). So I was expecting that if I could install a package with an underscore, I should also be able to import stuff from it and use it. This however did not work.

iatzak commented:
@IshaanDesai I tried again after deleting the project from TestPyPI. Unfortunately I got the same error:
![image](https://user-images.githubusercontent.com/112731474/204090860-a585243f-d70b-4c9b-b170-feaf12ddea33.png)

In [TestPyPI's help page](https://test.pypi.org/help/#file-name-reuse), it is mentioned

> PyPI will return these errors for one of these reasons:
>
> - Filename has been used and file exists
> - Filename has been used but file no longer exists
> - A file with the exact same content exists
>
> PyPI does not allow for a filename to be reused, even once a project has been deleted and recreated.

i.e., excluded projects' names are also accounted for. However, in my understanding, this text only covers exactly equal names, so I might need to take the time to look at each of my files individually (or redo the project from scratch). Until when would it be acceptable for me to solve this?

IshaanDesai commented:
> Until when would it be acceptable for me to solve this?

The original comment where I pointed out that the package was not working was meant more as a "for your information" and not necessarily something that you need to fix. The main goal here is that I try to give you the user side experience of a code you packaged. It is good that you deleted the project and tried to redo the whole thing. I did not know that TestPyPI considers names of packages that have been deleted while looking for possible name conflicts. This is already a good insight from you.

I would suggest that you note down the things you learnt here and remember them the next time you package a code. No need to fully fix this particular package.
